File: neetcode/go/_020/main.go
package main

import "fmt"

func main() {
	fmt.Println(isValid("()"))
	fmt.Println(isValid("()[]{}"))
	fmt.Println(isValid("(]"))
	fmt.Println(isValid("([])"))
}

func isValid(s string) bool {
	pairs := map[byte]byte{
		')': '(',
		'}': '{',
		']': '[',
	}

	stack := make([]byte, 0)
	for i := range s {
		open, exist := pairs[s[i]]
		if !exist {
			stack = append(stack, s[i])
			continue
		}
		if len(stack) == 0 {
			return false
		}
		if stack[len(stack)-1] != open {
			return false
		}

		stack = stack[:len(stack)-1]
	}

	return len(stack) == 0
}

File: neetcode/python/_020/main.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        mp = {'}': '{', ']': '[', ')': '(', '(': None, '[': None, '{': None}
        for char in s:
            n = len(stack)
            if n > 0 and mp[char] == stack[n-1]:
                stack = stack[:n-1]
            else:
                stack.append(char)
        return len(stack) == 0

print(Solution().isValid("()"))
print(Solution().isValid("()[]{}"))
print(Solution().isValid("(]"))
print(Solution().isValid("([])"))